### PR TITLE
Nano: Support best result for each trial in AutoML

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

Now NanoHPO selects model based on the last epoch's metric (for example, last epoch's val loss). However, it's more reasonable to select the one with the best metric.

### 1. Why the change?

https://github.com/intel-analytics/BigDL/issues/6760

### 2. User API changes

New parameter `mode` for `trainer.search`:
```python
model = CustomModel(
    out_dim = space.Categorical(16,32),
    dropout = space.Categorical(0.1,0.2),
    learning_rate = space.Real(0.001,0.01,log=True),
    batch_size = space.Categorical(32,64)
)

trainer = Trainer(
    logger=True,
    checkpoint_callback=False,
    max_epochs=2,
    use_hpo=True,
)

best_model = trainer.search(
    model,
    target_metric='val_loss',
    direction='minimize',
    n_trials=4,
    max_epochs=3,
    mode='best',  # default value is best, you can change to last.
)
```

### 3. Summary of the change 

- support obtain the best metric value for each trial
- provide two modes in `trainer.search`
- related ut

### 4. How to test?
- [x] Unit test
- [x] Application test

### CODE REVIEW

**Comment by rnwang04** [comment]
> others LGTM. Just one question, so are we now still using the checkpoint of the last epoch or we also switch to the checkpoint of the best epoch

Actually we don't save any checkpoint during AutoML, and if you mean the behavior of checkpoint in fitting, it depends on user themselves. Or you mean in chronos?
